File: src/margrete_rpc/_framing.py
from __future__ import annotations

import struct

from margrete_rpc._proto.margrete.rpc.v1 import messages_pb2
from margrete_rpc.errors import MargreteProtocolError

MAX_FRAME_SIZE = 16 * 1024 * 1024


def encode_frame(envelope: messages_pb2.Envelope) -> bytes:
    payload = envelope.SerializeToString()
    if len(payload) > MAX_FRAME_SIZE:
        raise MargreteProtocolError(f"payload too large: {len(payload)} bytes")
    return struct.pack("<I", len(payload)) + payload


def decode_frame(frame: bytes) -> messages_pb2.Envelope:
    if len(frame) < 4:
        raise MargreteProtocolError("frame header is truncated")
    size = struct.unpack("<I", frame[:4])[0]
    if size > MAX_FRAME_SIZE:
        raise MargreteProtocolError(f"frame too large: {size} bytes")
    payload = frame[4:]
    if len(payload) != size:
        raise MargreteProtocolError("frame payload is truncated")
    envelope = messages_pb2.Envelope()
    envelope.ParseFromString(payload)
    return envelope


__all__ = ["MAX_FRAME_SIZE", "decode_frame", "encode_frame"]

File: plugin/tests/framing.rs
use margrete_rpc::rpc::framing::{self, MAX_FRAME_SIZE};
use margrete_rpc::rpc::proto::{envelope, Envelope, PingRequest};

#[test]
fn frame_protocol_round_trips_envelope() {
    let envelope = Envelope {
        request_id: 42,
        body: Some(envelope::Body::PingRequest(PingRequest {})),
    };
    let frame = framing::encode(&envelope).unwrap();
    let decoded = framing::decode(&frame).unwrap();
    assert_eq!(decoded.request_id, 42);
    assert!(matches!(decoded.body, Some(envelope::Body::PingRequest(_))));
}

#[test]
fn frame_protocol_rejects_truncated_payload() {
    let frame = [4, 0, 0, 0, b'a'];
    let err = framing::decode(&frame).unwrap_err().to_string();
    assert!(err.contains("frame payload is truncated"));
}

#[test]
fn frame_protocol_round_trips_empty_payload() {
    let frame = [0, 0, 0, 0];
    let decoded = framing::decode(&frame).unwrap();
    assert_eq!(decoded.request_id, 0);
    assert!(!matches!(
        decoded.body,
        Some(envelope::Body::PingRequest(_))
    ));
}

#[test]
fn frame_protocol_rejects_oversized_header() {
    let too_big = MAX_FRAME_SIZE + 1;
    let frame = too_big.to_le_bytes();
    let err = framing::decode(&frame).unwrap_err().to_string();
    assert!(err.contains("frame payload is too large"));
}

#[test]
fn frame_protocol_rejects_invalid_protobuf_payload() {
    let frame = [1, 0, 0, 0, 0x08];
    let err = framing::decode(&frame).unwrap_err().to_string();
    assert!(err.contains("not a valid protobuf envelope"));
}
